src/corehost/cli/ijwhost/ijwthunk.cpp

```diff
@@ -0,0 +1,192 @@
+// Copyright (c) .NET Foundation and contributors. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+#include "ijwhost.h"
+#include "bootstrap_thunk_chunk.h"
+#include "corhdr.h"
+#include "error_codes.h"
+#include "trace.h"
+#include <heapapi.h>
+#include <new>
+#include <mutex>
+
+#ifdef _WIN64
+#define COR_VTABLE_PTRSIZED     COR_VTABLE_64BIT
+#define COR_VTABLE_NOT_PTRSIZED COR_VTABLE_32BIT
+#else
+#define COR_VTABLE_PTRSIZED     COR_VTABLE_32BIT
+#define COR_VTABLE_NOT_PTRSIZED COR_VTABLE_64BIT
+#endif
+
+namespace
+{
+    std::mutex g_thunkChunkLock{};
+    bootstrap_thunk_chunk* g_pVtableBootstrapThunkChunkList = nullptr;
+}
+
+HANDLE g_heapHandle;
+
+bool patch_vtable_entries(PEDecoder& pe)
+{
+    if (pe.IsILOnly())
+    {
+        // Nothing to do if the PE is IL-only.
+        return true;
+    }
+    
+    size_t numFixupRecords;
+    IMAGE_COR_VTABLEFIXUP* pFixupTable = pe.GetVTableFixups(&numFixupRecords);
+
+    if (numFixupRecords == 0)
+    {
+        // If we have no fixups, no need to allocate thunks.
+        return true;
+    }
+
+    size_t numThunks = 0;
+    for (size_t i = 0; i < numFixupRecords; ++i)
+    {
+        numThunks += pFixupTable[i].Count;
+    }
+
+    size_t chunkSize = sizeof(bootstrap_thunk_chunk) + sizeof(bootstrap_thunk) * numThunks;
+
+    void* pbChunk = HeapAlloc(g_heapHandle, 0, chunkSize);
+
+    if (pbChunk == nullptr)
```

Comment:
I want to talk about this. This will occur if the process runs out memory, so we return `false` and now what? Do we have a recourse here? My concern here is if we can't allocate the vtable for fixups, the function will return and then eventually the CLR will try to do something. I am more of the opinion that at IJW startup, we should A/V instead of attempting to handle an OOM.

Reply:
This function is called as part of the native module load, so if we fail to allocate, then we'll return false from `_CorDllMain` and tell Windows that we failed to load the module successfully. As a result, this module won't be successfully loaded into the process, let alone the CLR if we fail to allocate here.

Reply:
>As a result, this module won't be successfully loaded into the process, let alone the CLR if we fail to allocate here.

That is a good thing for sure. However, the question remains regarding how we or the user can proceed? Does this manifest itself as "module failed to load?" What is the developer experience here?

Reply:
Yep this manifests as a load failure, and if loading via `Assembly.Load` and family, it throws an exception just like if `Assembly.Load` failed on an IL-only assembly load.

Reply:
Sounds good from the managed side, but how about the native side? If I am in native and I make a call into a managed function, how does this come back to native? Can I assume the exception is marshalled as a structured exception or a C++ exception? I am concerned about crashes customers observe in their products and how well it can be diagnosed from a dump or log. I will say if an exception is thrown more than likely the process is coming down - so that is enough for me as long as that is true.